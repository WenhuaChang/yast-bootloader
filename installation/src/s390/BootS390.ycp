/**
 * File:
 *	BootS390.ycp
 *
 * Module:
 *	BootS390
 *
 * Summary:
 *	functions for aboot setup
 *
 * Authors:
 *	Klaus Kaempf <kkaempf@suse.de>
 *
 * $Id$
 *
 */
{
    module "BootS390";

    import "Boot";
    import "Installation";
    import "StorageControllers";

    textdomain "bootloader";

    /**
     * Write()
     *
     */

    global define boolean Write ()
    ``{
        Boot::Set();
	y2milestone( "install.inf dir:%1", SCR::Dir(.etc.install_inf));

	string root_device	= Boot::RootPartitionDevice;

	string zipl_conf = "# Generated by YaST2
[defaultboot]
default=ipl

[ipl]
target=/boot/zipl
image=/boot/kernel/image
ramdisk=/boot/initrd
parameters=\"" + StorageControllers::dasdParam + " root=" + root_device;

	zipl_conf = zipl_conf + "\"

[dumpdasd]
target=/boot/zipl
dumpto=/dev/dasd??

[dumptape]
target=/boot/zipl
dumpto=/dev/rtibm0
";

	SCR::Write (.target.string, "/etc/zipl.conf", zipl_conf);

	SCR::Execute (.target.bash, "/sbin/mk_initrd");

	// append chandev to /etc/chandev.conf
	string chandev = SCR::Read(.etc.install_inf.Chandev);
	if( chandev == nil )
	    {
	    chandev = "";
	    }
	if (chandev != "")
	{
	    SCR::Execute (.target.bash, "/bin/echo '"+chandev+"' >> /etc/chandev.conf");
	}
	integer zipl = SCR::Execute (.target.bash, "/sbin/zipl > /var/log/inst-log.zipl 2>&1");
	if (0 != zipl)
	{
	    string output = SCR::Read (.target.string, "/var/log/inst-log.zipl");
	    UI::OpenDialog ( `VBox (
		// label
		`Label (`id (`l1), sformat(_("/sbin/zipl returned exit code %1"), zipl)),
		`VSpacing (1),
		// label, ruchtext headline
		`Label (`id (`l2), _("Log contents:")),
		`RichText (`id (`out),  `opt(`plainText), output),
		`VSpacing (1),
		// pushbutton
		`PushButton (`id (`ok), _("&Continue"))
	    ));
	    UI::UserInput ();
	    UI::CloseDialog ();
	    return false;
	}
	return true;
    }
}
