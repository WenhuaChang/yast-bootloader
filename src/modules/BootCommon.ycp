/**
 * File:
 *      modules/BootCommon.ycp
 *
 * Module:
 *      Bootloader installation and configuration
 *
 * Summary:
 *      Data to be shared between common and bootloader-specific parts of
 *      bootloader configurator/installator, generic versions of bootloader
 *      specific functions
 *
 * Authors:
 *      Jiri Srain <jsrain@suse.cz>
 *
 * $Id$
 *
 */

{
    module "BootCommon";

    import "Arch";

    import "Storage";

    textdomain "bootloader";

    include "bootloader/routines/misc.ycp";
    include "bootloader/routines/widgets.ycp";
    include "bootloader/routines/helps.ycp";
    include "bootloader/routines/lilolike.ycp";

// common variables

    /**
      * type of bootloader to configuer/being configured
      * shall be one of "lilo", "grub", "silo", "milo", "aboot",
      * "elilo", "ppc", "s390", "mips"
      */
    string loader_type = nil;

// sysconfig variables

    /**
      * location description
      * on i386 can be mbr, root, boot, floppy, custom, none
      * other architectures may differ
      */
    global string location = "";

    /**
      * Device to install bootloader to
      */
    global string loader_device = "";

// variables for storing data

    /**
      * list of section
      */
    global list sections = [];

    /**
      * map of global options and values
      */
    global map globals = $[];

    /**
      * string sepresenting device name of /boot partition
      * same as RootPartitionDevice if no separate /boot partition
      */
    global string BootPartitionDevice = "";

    /**
      * string representing device name of / partition
      */
    global string RootPartitionDevice = "";
    /**
      * device holding MBR for bootloader
      */
    global string mbrDisk = "";

// state variables

    /**
      * was the propose function called (== true)
      */
    global boolean was_proposed = false;
    /**
      * Were module settings read (== true)
      */
    global boolean was_read = false;
    /**
      * were settings changed (== true)
      */
    global boolean changed = false;
    /**
      * Was bootloader location changed? (== true)
      */
    global boolean location_changed = false;
    /**
      * Were configuration files manually edited and chamged?
      */
    global boolean files_edited = false;
    /**
      * Has been files edited warning already shown?
      */
    global boolean files_edited_warned = false;
    /**
      * Shall be settings saved when finishing bootloader configuration?
      */
    global boolean save_on_finish = true;
    /**
      * time of last change of partitioning
      */
    global integer partitioning_last_change = 0;

// saving mode setting functions

    /**
      * map of save mode settings
      */
    global map write_settings = $[];

// summary dialog state

    /**
      * Show verbose summary output
      */
    global boolean verbose = false;

// other variables

    /**
      * Settings of other bootloaders used when switching bootloader
      */
    global map other_bl = $[];

    /**
      * Option types for differnt bootloaders
      */
    global map opt_types = $[];

    /**
      * Map for temporary state storing
      */
    global map tmp_store = $[];

    /**
      * Map to remember what settings were changed
      */
    global map changed_settings = $[];
    /**
      * Kernel parameters at previous detection
      */
    global string kernelCmdLine = "";

// bootloader installation variables

    /**
      * Replace MBR with generic code after bootloader installation?
      */
    global boolean repl_mbr = false;

    /**
      * Activate bootloader partition during installation?
      */
    global boolean activate = true;

// generic versions of bootloader-specific functions

    /**
      * Export bootloader settings to a map
      * @return bootloader settings
      */
    global define map Export () ``{
        map exp = $[
		"global": globals,
		"sections" : sections,
	    ];
	return exp;
    }

    /**
      * Import settings from a map
      * @param settigns map of bootloader settings
      */
    global define void Import (map settings) ``{
        globals = lookup(settings, "global", $[]);
        sections  = lookup(settings, "sections", []);
	return;
    }
    /**
      * Read settings from disk
      * @return boolean true on success
      */
    global define boolean Read () ``{
	loadCfgFileFromAgent ();
	loader_device = globals["boot"]:"/dev/null";
        return true;
    }

    /**
      * Reset bootloader settings
      */
    global define void Reset () ``{
        sections = [];
        globals = $[];
//	DetectDisks ();
	repl_mbr = false;
	activate = true;
    }

    /**
      * Propose bootloader settings
      */
    global define void Propose () ``{
	y2error ("No generic propose function available");
    }
    /**
      * Save all bootloader configuration files
      * @return boolean true if success
      */
    global define boolean Save () ``{
	string bl = getLoaderType (false);
	saveCfgFileToAgent ();
	SCR::Write(list2path([bl]), nil);
	return true;
    }
    /**
      * Display bootloader summary
      * @return list of summary lines
      */
    global define list(string) Summary () ``{
	string default = BootCommon::getLoaderType (false) == "grub"
	    ? sections[globals["default"]:0, "title"]:sections[0, "title"]:""
	    : globals["default"]:sections[0, "label"]:"";
	map targetMap = Storage::GetTargetMap ();
	map boot_target = targetMap[loader_device]:$[];
	string target_name = "";
	if (boot_target == $[])
	    target_name = loader_device;
	else
	target_name = boot_target["name"]:"disk";

	list(string) result = [];
	result = add (result, sformat (_("Bootloader type: %1"),
		getLoaderName (getLoaderType (false))));
	result = add (result, sformat (_("Location: %1"), target_name));
	list(string) sects = [];
	foreach (`s, sections, ``{
	    string title = s[
		getLoaderType (false) == "grub" ? "title" : "label"
	    ]:"";
	    string def = title == default ? _(" (default)") : "";
	    sects = add (sects, sformat ("%1%2", title, def));
	});
	result = add (result, sformat (_("Sections: %1"),
	    mergestring (sects, ", ")));
	if (loader_device == "/dev/null")
	    result = add (result, _("Don't install bootloader, just create
configuration files"));
	return result;
    }

    /**
      * Update read settings to new version of configuration files
      */
    global define void Update () ``{
    }

    /**
      * Write bootloader settings to disk
      * @return boolean true on success
      */
    global define boolean Write () ``{
	y2error ("No generic write possible");
	return false;
    }

    /**
      * Get map of bootloader options
      * @return map of options
      */
    global define map GetOptions () ``{
	map opts = $[
	    `global: [
	        "append",
	        "backup",
	        "boot",
	        "change-rules",
	        "compact",
	        "default",
	        "delay",
	        "disk",
	        "disktab",
	        "fix-table",
		"force-backup",
	        "ignore-table",
	        "install",
	        "lba32",
	        "linear",
	        "lock",
	        "map",
	        "menu-title",
	        "menu-scheme",
	        "message",
	        "nowarn",
	        "optional",
	        "password",
	        "prompt",
	        "restricted",
	        "serial",
	        "timeout",
	        "verbose",
	        "vga",
	    ],
	    "image" : [
	        "append",
	        "initrd",
	        "literal",
	        "ramdisk",
	        "read-only",
	        "read-write",
	        "root",
	        "vga",
	        "label",
	        "alias",
	        "lock",
	        "optional",
	        "password",
	        "restricted",
	    ],
	    "other" : [
	        "loader",
	        "table",
	        "change",
	        "map-drive",
	        "unsafe",
	        "label",
	        "alias",
	        "lock",
	        "optional",
	        "password",
	        "restricted",
	    ]
	];
	return opts;
    }

// end of generic versions of bootloader-specific functions
//-----------------------------------------------------------------------------

// functions for fetching settings to tmp map and storing back

    /**
      * Fetch settings to temporary variables
      * @return sumbol wizard sequencer symbol
      */
    global define symbol fetchSettings () ``{
	tmp_store = $[];
	tmp_store["globals"] = eval(globals);
	tmp_store["sections"] = eval(sections);
	tmp_store["location"] = location;
	tmp_store["loader_device"] = loader_device;
	tmp_store["activate"] = activate;
	tmp_store["repl_mbr"] = repl_mbr;
	return `next;
    }

    /**
      * Store settings from temporary variables to globals
      * @return wizard sequencer symbol
      */
    global define symbol storeSettings () ``{
	globals = tmp_store["globals"]:$[];
	sections = tmp_store["sections"]:[];
	location = tmp_store["location"]:"";
	loader_device = tmp_store["loader_device"]:"";
	loader_device = GetBootloaderDevice ();
	repl_mbr = tmp_store["repl_mbr"]:false;
	activate = tmp_store["activate"]:false;
	return `next;
    }

//-----------------------------------------------------------------------------
// common functions start

// bootloader type handling functions

    /**
      * Get currently used bootloader, detect if not set yet
      * @param recheck boolean force checking bootloader
      * @return string botloader type
      */
    global define string getLoaderType (boolean recheck) ``{
	if ((! recheck) && (loader_type != nil))
	    return loader_type;
	// read bootloader to use from disk
	if (Mode::update || Mode::normal)
	{
	    loader_type = SCR::Read (.sysconfig.bootloader.LOADER_TYPE);
	    if (loader_type != nil && loader_type != "")
	    {
		y2milestone ("Sysconfig bootloader is %1, using", loader_type);
		return loader_type;
	    }
	}
	// detect bootloader
	loader_type = SCR::Read (.probe.boot_arch);
	y2milestone ("Bootloader detection returned %1", loader_type);
	if (loader_type == "grub" && Storage::UseLilo())
	    loader_type = "lilo";

	// this shall install packages if possible
	setLoaderType (loader_type);

	// this was wrong - storage, which initializes before package maneger,
	// needs calls this function
	// check bootloader package to be installed
/*	if ((Arch::i386 || Arch::x86_64) && (Mode::initial)){
	    if (loader_type == "grub" && ! Pkg::IsSelected ("grub"))
		loader_type = "lilo";
            if (loader_type == "lilo" && ! Pkg::IsSelected ("lilo"))
                loader_type = "grub";
	    if (loader_type == "grub" && ! Pkg::IsSelected ("grub"))
	    {
		// no bootloader is installed
		noBootloaderPopup ();
	    }
	}*/
	y2milestone ("Detected bootloader %1", loader_type);
	return loader_type;
    }
    /**
      * set type of bootloader
      * @param bootloader string type of bootloader
      */
    global define void setLoaderType (string bootloader) ``{
	if (contains(BootCommon::bootloaders, bootloader)
	    && ! Mode::test)
	{
	    if (Mode::normal)
	    {
		boolean installed = false;
		foreach (`p, getBootloaderPackages (bootloader), ``{
		    if (! Pkg::IsProvided (bootloader))
		    {
		        Pkg::PkgInstall (bootloader);
			installed = false;
		    }
		});
		if (installed)
		{
		    Pkg::PkgSolve ();
		    Pkg::PkgCommit (0);
		    WFM::CallModule("inst_suseconfig", []);
		}
	    }
	    else
	    {/* FIXME some checking must be added, this idea might be correct
		import "ForcedPackages";
		foreach (`b, bootloaders, ``{
		    foreach (p, getBootloaderPackages (b), ``{
			ForcedPackages::RemoveFromForceInstall (b);
		    });
		});
		foreach (`p, getBootloaderPackages (bootloader), ``{
		    ForcedPackages::RemoveFromForceNotInstall (p);
		    ForcedPackages::ForceInstall (p);
		});
	*/
	    }
	}
	loader_type = bootloader;
    }
    /**
      * List bootloaders available for configured architecture
      * @return list of bootloaders
      */
    global define list getBootloaders () ``{
	list ret = [getLoaderType (false)];
	if (Arch::i386)
	{
	    ret = merge (ret, ["lilo", "grub"]);
	}
	return toset (ret);
    }

}


