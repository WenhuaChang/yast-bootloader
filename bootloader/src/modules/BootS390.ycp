/**
 * File:
 *      modules/BootS390.ycp
 *
 * Module:
 *      Bootloader installation and configuration
 *
 * Summary:
 *      Module containing specific functions for S390 configuration
 *      and installation
 *
 * Authors:
 *      Jiri Srain <jsrain@suse.cz>
 *
 * $Id$
 *
 */

{
    module "BootS390";

    textdomain "bootloader";

    import "Kernel";
    import "StorageControllers";
    import "BootCommon";
    import "Mode";
    import "Stage";
    import "Installation";

    include "bootloader/routines/popups.ycp";
    include "bootloader/routines/routines.ycp";

// local data


    map<string,map<string,any> > hw_descr = $[
	"ctc" : $[
	    "skeleton": "hwcfg-ctc",
	    "target" : "ctc-bus-ccw-%1",
	    "options" : $[
		"CCW_CHAN_IDS" : "%1 %2",
		"CCW_CHAN_MODE" : "%3",
	    ],
	],
	"qeth" : $[
	    "skeleton" : "hwcfg-qeth",
	    "target" : "qeth-bus-ccw-%1",
	    "options" : $[
		"CCW_CHAN_IDS" : "%1 %2 %3",
		"CCW_CHAN_MODE" : "%4",
	    ],
	],
	"iucv" : $[
	    "skeleton" : "hwcfg-iucv",
	    "target" : "iucv-id-%1",
	],
    ];

// very simple UI - only select loader type

    /**
      * Create a list of options for the table
      * @param descr map description of the table
      * @return list of option ids as it should be displayed
      */
    global define list getTableContents (map descr) ``{
	return ["__loader_type"];
    }

    /**
      * Get map of widget
      * @return map of widget
      */
    global define map<string,any> getGlobalSectionWidget () ``{
	map<string,any> ret = BootCommon::getGlobalSectionWidget ();
	ret["ids"] = getTableContents;
	return ret;
    }

    /**
      * Get map of dialogs and widgets
      * @return map of dialogs and widgets
      */
    global define map<string,map<string,any> > GetWidgetMaps () ``{
	return $[
	    "globalsection" : getGlobalSectionWidget (),
	];
    }



// misc. functions

    /**
      * Update /etc/zip.conf before the update (fix kernel path)
      */
    global define void UpdateZiplConfBeforeUpdate () {
	// update the partition layout if needed
	// ?? FIXME any other conditions?
	if (Mode::update ())
	{
	    map output = (map)SCR::Execute (.target.bash_output, sformat (
	      "/bin/cp %1/etc/zipl.conf %1/etc/zipl.conf.YaST2save",
	      Installation::destdir));
	    y2milestone ("Result of backup original elilo.conf: %1", output);
	    output = (map)SCR::Execute (.target.bash_output, sformat (
		"/bin/cp /usr/lib/YaST2/bin/update_zipl_conf %1/usr/lib/YaST2/bin/",
		Installation::destdir
	    ));
	    y2milestone ("Result of copying from / to %1: %2",
		Installation::destdir,
		output
	    );
	    output = (map)SCR::Execute (.target.bash_output, sformat (
		"chroot %1 /usr/lib/YaST2/bin/update_zipl_conf",
		Installation::destdir));
	    if (output["exit"]:0 != 0)
	    {
		y2error ("Update of /etc/zipl.conf failed: %1", output);
	    }
	    else
	    {
		BootCommon::was_proposed = false;
		BootCommon::was_read = false;
	    }
	}
    }

    /**
      * Empty summary dialog
      * @return symbol always `back
      */
    global define symbol runSummaryDialog () ``{
	displayNoSupportPopup ();
	return `back;
    }

    /**
      * Update /etc/sysconfig/hardware configuration
      * Use data from install.inf file
      * @return boolean true on success
      */
    global define boolean updateHardwareConfig () ``{
	if (! Stage::initial () || Mode::update ())
	    return true;

	boolean failed = false;
	string cfg = (string)SCR::Read (.etc.install_inf.Hardware);
	y2milestone ("Read hardware configuration from install.inf: %1", cfg);
	list<string> l = splitstring (cfg, ";");
	foreach (string s, l, ``{
	    list<string> args = splitstring (s, ",");
	    args = maplist (string a, args, ``{
		while (a != "" && substring (a, 0, 1) == " ")
		    a = substring (a, 1);
		while (a != "" && substring (a, size (a) - 1, 1) == " ")
		    a = substring (a, 0, size (a) - 1);
		return a;
	    });
	    string key = args[0]:"";
	    string a1 = args[1]:"";
	    string a2 = args[2]:"";
	    string a3 = args[3]:"";
	    string a4 = args[4]:"";
	    if (key != "")
	    {
		map<string,any> descr = hw_descr[key]:$[];
		string src = (string)(descr["skeleton"]:"");
		string dst = sformat ((string)(descr["target"]:""),
		    a1, a2, a3, a4);
		y2milestone ("Template: %1, Target: %2", src, dst);
		string command = sformat (
    "/bin/cp /etc/sysconfig/hardware/skel/%1 /etc/sysconfig/hardware/hwcfg-%2",
		    src, dst);
		if (0 != SCR::Execute (.target.bash, command))
		{
		    Report::Error (
		        // error report
			_("Copying hardware configuration template failed."));
		    failed = true;
		}
		path p = add (.sysconfig.hardware.value, dst);
		foreach (string k, string v,
		    (map<string,string>)(descr["options"]:$[]),
		``{
		    path op = add (p, k);
		    v = sformat (v, a1, a2, a3, a4);
		    if (! SCR::Write (op, v))
			failed = true;
		});
	    }
	});
	if (! SCR::Write (.sysconfig.hardware, nil))
	    failed = true;;
	return failed;
    }


// general functions

    /**
      * Propose bootloader settings
      */
    global define void Propose () ``{
	BootCommon::DetectDisks ();
	string root_device = BootCommon::RootPartitionDevice;
	string parameters = sformat ("%1 root=%2 %3 %4 %5",
	    StorageControllers::dasdParam,
	    root_device,
	    BootCommon::GetAdditionalKernelParams (),
	    "selinux=0",
	    "TERM=dumb");

	BootCommon::globals = [];
	BootCommon::sections = [
	    [
		$[ "key" : "label", "value" : "defaultboot" ],
		$[ "key" : "default", "value" : "ipl" ],
	    ],
	    [
		$[ "key" : "label", "value" : "ipl" ],
		$[ "key" : "target", "value" : "/boot/zipl" ],
		$[ "key" : "image",
		   "value" : sformat ("/boot/%1", Kernel::GetBinary ()) ],
		$[ "key" : "ramdisk", "value" : "/boot/initrd" ],
		$[ "key" : "parameters", "value" : parameters ],
	    ],
	    [
		$[ "key" : "label", "value" : "dumpdasd" ],
		$[ "key" : "target", "value" : "/boot/zipl" ],
		$[ "key" : "dumpto", "value" : "/dev/dasd??" ],
	    ],
	    [
		$[ "key" : "label", "value" : "dumptape" ],
		$[ "key" : "target", "value" : "/boot/zipl" ],
		$[ "key" : "dumpto", "value" : "/dev/rtibm0" ],
	    ],
	];
    }

    /**
      * Read settings from disk
      * @return boolean true on success
      */
    global define boolean Read () ``{
	BootCommon::DetectDisks ();
	BootCommon::changed = true;
	return BootCommon::Read ();
    }

    /**
      * Save all bootloader configuration files
      * @return boolean true if success
      */
    global define boolean Save () ``{
	y2milestone ("Saving S390 cfg. file");
	boolean ret = BootCommon::Save ();

	integer exit = (integer)SCR::Execute (.target.bash,
	    "test -d /boot/zipl || mkdir -p /boot/zipl");
	ret =  ret && (exit == 0);

	if (Mode::normal ())
	    return ret;

	updateHardwareConfig ();
// chandev is obsolete, syscofnig/hardware is used instead
/*	// append chandev to /etc/chandev.conf
	string chandev = (string)SCR::Read(.etc.install_inf.Chandev);
	if( chandev == nil )
	{
	    chandev = "";
	}

	if (chandev != ""
	    && (SCR::Read (.target.size, "/etc/chandev.conf") == -1))
	{
	    ret = ret && (0 == SCR::Execute (.target.bash,
		"/bin/echo '"+chandev+"' >> /etc/chandev.conf"));
	}*/
	return ret;
    }

    /**
      * Display bootloader summary
      * @return a list of summary lines
      */
    global define list<string> Summary () ``{
	// summary
	return [_("Install S390 Boot Loader")];
    }

    /**
      * Update read settings to new version of configuration files
      */
    global define void Update () ``{
	y2milestone ("No update functionality implemented");
    }

    /**
      * Write bootloader settings to disk
      * @return boolean true on success
      */
    global define boolean Write () ``{
	string bl_command = "/sbin/zipl > /var/log/YaST2/y2logzipl 2>&1";
	string bl_logfile = "/var/log/YaST2/y2logzipl";
	return BootCommon::installBootLoader (bl_command, bl_logfile);
    }

    /**
      * Update bootlaoder configuration sequence
      * @param ws_data map of wizard sequencer data
      * @return map modified map of wizard sequencer data
      */
    global define map FixSequence (map ws_data) ``{
        map aliases = $[
	    "summary" : ``(BootS390::runSummaryDialog ()),
        ];
        ws_data["aliases"] = aliases;
        return ws_data;
    }

    /**
      * Get the list of all dump devices
      * @return list of all dump devices, as list of maps containg keys
      *  "label", "target" and "dumpto"
      */
    global define list<map<string,string> > GetDumpDevices () ``{
	list<map<string,string> > dd = maplist (list<map> d,
	    BootCommon::sections,
	``(
	    listmap (map e, d, ``(
		$[ e["key"]:"": e["value"]:"" ]
	    ))
	));
	return filter (map<string,string> s, dd, ``(
	    s["dumpto"]:"" != ""
	));
    }

    /**
      * Set the list of all dump devices
      * @param dump_devices list in the same form as result of GetDumpDevices ()
      */
    global define void SetDumpDevices (list<map<string,string> > dump_devices) {
	BootCommon::sections = filter (list<map> s, BootCommon::sections, ``(
	    lookup_value (s, "dumpto", "") == ""
	));
	dump_devices = maplist (map<string,string> d, dump_devices, ``{
	    if (d["dumpto"]:"" == "")
		return nil;
	    if (d["target"]:"" == "")
		d["target"] = "/boot/zipl";
	    if (d["label"]:"" == "")
		d["label"] = "dump" + filterchars (d["dumpto"]:"", "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789");
	    return d;
	});
	dump_devices = filter (map<string,string> d, dump_devices,``(d != nil));
	list<list<map> > dd = maplist (map<string,string> d, dump_devices, ``(
	    (list<map>)maplist (string k, string v, d, ``(
		$[ "key" : k, "value" : v ]
	    ))
	));
	BootCommon::sections = (list<list<map> >)merge (BootCommon::sections, dd);
    }

    /**
      * Return map of provided functions
      * @return map map of functions (eg. $["write":``(BootS390::Write ())])
      */
    global define map<string, any> GetFunctions () ``{
	return $[
	    "read"	: BootS390::Read,
	    "propose"	: BootS390::Propose,
	    "save"	: BootS390::Save,
	    "summary"	: BootS390::Summary,
	    "update"	: BootS390::Update,
	    "write"	: BootS390::Write,
//	    "sequence"	: BootS390::FixSequence,
	];
    }

    /**
      * Initializer of S390 bootloader
      */
    global define void Initializer () ``{
	y2milestone ("Called S390 (zipl) initializer");
	BootCommon::current_bootloader_attribs = $[
	    "section_title" : "label",
	    "kernel_key" : "image",
	    "propose" : true,
	    "read" : true,
	    "scratch" : true,
	];
	// TODO no UI available yet
	BootCommon::cwm_options = (map<string, map<string,any> >)union (
	    BootCommon::GetOptionMaps (),
	    $[]);//BootS390::GetOptionMaps ());
	BootCommon::cwm_widgets = (map<string, map<string,any> >)union (
	    BootCommon::GetWidgetMaps (),
	    BootS390::GetWidgetMaps ());
    }


    /**
      * Constructor
      */
    global define void BootS390 () ``{
	BootCommon::bootloader_attribs["s390"] = $[
            "loader_name" : "zipl",
	    "required_packages" : [], // FIXME
	    "initializer" : BootS390::Initializer,
        ];
    }

}
