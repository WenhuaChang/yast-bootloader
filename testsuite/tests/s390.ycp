/**
 * File:
 *  grub.ycp
 *
 * Module:
 *  Bootloader configurator
 *
 * Summary:
 *  Bootloader GRUB misc functions testsuite
 *
 * Authors:
 *  Jiri Srain <jsrain@suse.cz>
 *
 * $Id$
 *
 */

{
    include "testsuite.ycp";

    map READ_I = $[
        "target" : $[
            "size" : -1,
            "tmpdir" : "/tmp/",
	    "yast2" : nil,
        ],
        "probe" : $[
            "architecture" : "i386",
            "has_apm" : true,
            "has_pcmcia" : false,
            "has_smp" : false,
            "system" : [],
            "memory" : [],
            "cpu" : [],
            "cdrom" : $[
                "manual" : [],
            ],
            "floppy" : $[
                "manual" : [],
            ],
        ],
        "sysconfig" : $[
            "console" : $[
                "CONSOLE_FONT" : "",
                "CONSOLE_SCREENMAP" : "",
                "CONSOLE_UNICODEMAP" : "",
                "CONSOLE_MAGIC" : "",
                "CONSOLE_ENCODING" : "",
            ],
            "language" : $[
                "RC_LANG" : "",
                "DEFAULT_LANGUAGE" : "",
            ],
        ],
        "etc" : $[
            "install_inf" : $[
                "Cmdline" : "",
            ],
        ],
        "proc" : $[
            "cpuinfo" : $[
                "value" : $[
                    "0" : $[
                        "flags" : ""
                    ],
                ],
            ],
        ]
    ];
    map WRITE_I = $[];
    map EXEC_I = $[
	"target" : $[
	    "bash_output" : $[],
	],
    ];

    TESTSUITE_INIT ([READ_I, WRITE_I, EXEC_I], 0);
    import "Mode";
    Mode::config = true; // yes, this is ugly, bus solves Product:: constructor
    import "BootS390";
    import "BootCommon";

        BootCommon::sections = [
            [
                $[ "key" : "label", "value" : "defaultboot" ],
                $[ "key" : "default", "value" : "ipl" ],
            ],
            [
                $[ "key" : "label", "value" : "ipl" ],
                $[ "key" : "target", "value" : "/boot/zipl" ],
                $[ "key" : "image", "value" : "/boot/kernel/image" ],
                $[ "key" : "ramdisk", "value" : "/boot/initrd" ],
                $[ "key" : "parameters", "value" : "parameters" ],
            ],
            [
                $[ "key" : "label", "value" : "dumpdasd" ],
                $[ "key" : "target", "value" : "/boot/zipl" ],
                $[ "key" : "dumpto", "value" : "/dev/dasd??" ],
            ],
            [
                $[ "key" : "label", "value" : "dumptape" ],
                $[ "key" : "target", "value" : "/boot/zipl" ],
                $[ "key" : "dumpto", "value" : "/dev/rtibm0" ],
            ],
        ];

    y2error ("%1", BootCommon::sections);

    DUMP ("======================================");

    list<map<string,string> > dumps = (list<map<string,string> >)
	TEST (``(BootS390::GetDumpDevices ()), [], nil);

    DUMP ("======================================");

    TEST (``(BootS390::SetDumpDevices (dumps)), [], nil);

    DUMP ("======================================");

    y2error ("%1", BootCommon::sections);

    DUMP ("======================================");

    TEST (``(BootS390::GetDumpDevices ()), [], nil);


}
