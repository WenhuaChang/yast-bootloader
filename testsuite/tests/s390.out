Read	.target.size "/etc/install.inf" -1
Read	.probe.architecture "i386"
Read	.probe.has_apm true
Read	.probe.has_pcmcia false
Read	.probe.is_uml false
Read	.probe.has_smp false
Read	.probe.system []
Read	.proc.cpuinfo.value."0"."flags" ""
Read	.probe.memory []
Read	.sysconfig.console.CONSOLE_ENCODING ""
Execute	.target.bash_output "locale -k charmap" $[]
Read	.target.tmpdir "/tmp/"
Log	[[$["key":"label", "value":"defaultboot"], $["key":"default", "value":"ipl"]], [$["key":"label", "value":"ipl"], $["key":"target", "value":"/boot/zipl"], $["key":"image", "value":"/boot/kernel/image"], $["key":"ramdisk", "value":"/boot/initrd"], $["key":"parameters", "value":"parameters"]], [$["key":"label", "value":"dumpdasd"], $["key":"target", "value":"/boot/zipl"], $["key":"dumpto", "value":"/dev/dasd??"]], [$["key":"label", "value":"dumptape"], $["key":"target", "value":"/boot/zipl"], $["key":"dumpto", "value":"/dev/rtibm0"]]]
Dump	======================================
Return	[$["dumpto":"/dev/dasd??", "label":"dumpdasd", "target":"/boot/zipl"], $["dumpto":"/dev/rtibm0", "label":"dumptape", "target":"/boot/zipl"]]
Dump	======================================
Return	nil
Dump	======================================
Log	[[$["key":"label", "value":"defaultboot"], $["key":"default", "value":"ipl"]], [$["key":"label", "value":"ipl"], $["key":"target", "value":"/boot/zipl"], $["key":"image", "value":"/boot/kernel/image"], $["key":"ramdisk", "value":"/boot/initrd"], $["key":"parameters", "value":"parameters"]], [$["key":"dumpto", "value":"/dev/dasd??"], $["key":"label", "value":"dumpdasd"], $["key":"target", "value":"/boot/zipl"]], [$["key":"dumpto", "value":"/dev/rtibm0"], $["key":"label", "value":"dumptape"], $["key":"target", "value":"/boot/zipl"]]]
Dump	======================================
Return	[$["dumpto":"/dev/dasd??", "label":"dumpdasd", "target":"/boot/zipl"], $["dumpto":"/dev/rtibm0", "label":"dumptape", "target":"/boot/zipl"]]
